[DOCS] Update Docker getting started for the Stack

lockewritesdocs opened the pull request:
This PR replaces the steps for [Running the Stack with Docker Compose](https://www.elastic.co/guide/en/elastic-stack-get-started/8.0/get-started-docker.html#run-stack-docker) with steps to start the Stack on Docker using auto-configured security. 

Preview link: https://stack-docs_1888.docs-preview.app.elstc.co/guide/en/elastic-stack-get-started/master/get-started-stack-docker.html

A follow-up PR will address updating the steps for _Running the Stack with Docker Compose_.

bytebilly commented on docs/en/getting-started/get-started-docker.asciidoc:
```diff
@@ -1,69 +1,177 @@
+
 endif::[]
+
+ifeval::["{release-state}"!="unreleased"]
```
Should we show the command anyway, like we do for Kibana in the following section?

bytebilly commented on docs/en/getting-started/get-started-docker.asciidoc:
```diff
@@ -1,69 +1,177 @@
 
-. Open {kib} to load sample data and interact with the cluster:
-http://localhost:5601.
+. In a new terminal session, start {kib} and connect it to your {es} container:
```
Should we add the warning in case the version is not released yet?

bytebilly commented on docs/en/getting-started/get-started-docker.asciidoc:
```diff
@@ -1,69 +1,177 @@
---------------------------------------------
+[source,sh,subs="attributes"]
+----
+docker run --name es-node01 -p 9200:9200 -p 9300:9300 -it docker.elastic.co/elasticsearch/elasticsearch:{version}
```
Do we really need to expose port 9300 to the world since we are creating a single-node cluster?

lockewritesdocs replied:
This is the approach we implemented in the docs for [installing Kibana on Docker](https://www.elastic.co/guide/en/kibana/8.0/docker.html#run-kibana-on-docker-for-dev). If we don't want to expose port `9300`, then I'll update it in both places.

bytebilly replied:
Thanks

bytebilly commented on docs/en/getting-started/get-started-docker.asciidoc:
```diff
@@ -1,69 +1,177 @@
+
+[source,sh]
+----
+docker exec -it es-node01 /usr/share/elasticsearch/bin/elasticsearch-reset-password
```
```suggestion
docker exec -it es-node01 /usr/share/elasticsearch/bin/elasticsearch-reset-password -u elastic
```

lockewritesdocs replied:
Thanks for pointing out this omission -- I'll update the command with the `-u` option across all of the Docker documentation.

bytebilly commented on docs/en/getting-started/get-started-docker.asciidoc:
```diff
@@ -1,69 +1,177 @@
---------------------------------------------
+[source,sh]
+----
+docker cp es-node01:/usr/share/elasticsearch/config/tls_auto_config_*/http_ca.crt .
```
This command is not working, as `*` cannot be used for path expansion from outside docker. There are a few changes that we can do to make it simpler, otherwise we should list folders first or find another way.

lockewritesdocs replied:
We could just list the command as `tls_auto_config_<timestamp>` and indicate that users need to modify that portion of their command to match the filename on disk. Open to alternatives if you have preferred options.

bytebilly commented on docs/en/getting-started/get-started-docker.asciidoc:
```diff
@@ -1,69 +1,177 @@
++
+[source,sh,subs="attributes"]
+----
+docker run --name kib-01 --net elastic -p 5601:5601 docker.elastic.co/kibana/kibana:{version}
```
This command fails because the `elastic` network doesn't exist.

lockewritesdocs replied:
I'll add `--net elastic` to the command for starting Elasticsearch on line 106.

bytebilly replied:
I'm not sure if this solves the problem, IIRC the network must be created explicitly.
Can we just use the default network instead?

bytebilly commented on docs/en/getting-started/get-started-docker.asciidoc:
```diff
@@ -1,69 +1,177 @@
++
+[source,sh,subs="attributes"]
+----
+docker run --name kib-01 --net elastic -p 5601:5601 docker.elastic.co/kibana/kibana:{version}
```
Container name should be consistent across Elasticsearch and Kibana, e.g. `kibana01` here or `es-01` there.

lockewritesdocs replied:
I'll change to `es-node01` throughout (and subsequently `es-node02` and `es-node01`) and `kib01`.

bytebilly commented on docs/en/getting-started/get-started-docker.asciidoc:
```diff
@@ -1,69 +1,177 @@
-trial license without setting up TLS, we advise securing your stack from the
-start.
+  .. Log in to {kib} as the `elastic` user with the password that was generated
+  when you started {es}.
```
We could add clean-up  instuctions to stop and remove containers after use.

lockewritesdocs replied:
++ I'll add a section for removing the containers.

bytebilly reviewed (changes requested):
@lockewritesdocs I left suggestions for improvements and I marked commands that fail if run as-is. Ping me if you need extra technical details.

lockewritesdocs commented:
@bytebilly -- I addressed your comments and made a few other improvements that I'll port to the other Docker docs. Let me know if additional changes are needed!

jkakavas commented on docs/en/getting-started/docker/docker-compose.yml:
```diff
@@ -1,13 +1,13 @@
-version: '2.2'
+version: "2.2"
```
IIUC, this has been optional for a while and the current version is 3.8. Do we use any features below that require us to stay in 2.2 ? Can we change to latest or remove this ? @bytebilly might know

bytebilly replied:
We are rewriting the `docker-compose.yml` file entirely, so we don't know yet.
Since this is for getting started, we should not assume users have the latest `docker-compose` and it would be nice to make it compatible with the lowest version possible.

It's better to make it explicit, since some of the keywords have been removed in version 3 but they still work if you specify version 2 in the file.

jkakavas commented on docs/en/getting-started/docker/docker-compose.yml:
```diff
@@ -1,13 +1,13 @@
     image: docker.elastic.co/elasticsearch/elasticsearch:{version}
-    container_name: es01
+    container_name: es-node01
     environment:
```
I suggest that if we're keeping this compose file around, we should make it obvious that TLS is not enabled, and do so by setting xpack.security.http.ssl.enabled=false and xpack.security.transport.ssl.enabled=false here. Autoconfiguration won't be triggered because of the other settings, but I suggest we make this explicit too by setting xpack.security.autoconfiguration.enabled=false

jkakavas replied:
I didn't read the PR description so this comment is most probably relevant for the follow up PR.

jkakavas commented on docs/en/getting-started/get-started-docker.asciidoc:
```diff
@@ -1,69 +1,269 @@
+
+You can also <<get-started-docker-tls,start the {stack} with Docker Compose>> to
+create a secured, multi-node cluster with a connected {kib} instance. This
+option requires more manual steps, but results in a more robust cluster.
```
Should we qualify what we mean with "robust cluster" here ? It reads slightly generic

lockewritesdocs replied:
Yeah, I think that we can expand a bit more without getting too in depth. How about:
>This option requires more manual steps, but results in a more resilient cluster with greater capacity and reliability.

jkakavas commented on docs/en/getting-started/get-started-docker.asciidoc:
```diff
@@ -1,69 +1,269 @@
+[[run-docker-secure]]
+[discrete]
+=== Start {es} and {kib} in Docker
+Starting in {es} 8.0, security is enabled by default. With security enabled,
+{stack} {security-features} require TLS encryption for the transport networking
+layer, or your cluster will fail to start.
```
The TLS comment makes me think as a reader that I need to so something extra because security is enabled. I think this wording made sense in 7.x but maybe not so now ? It is true, but TLS will be enable by default if users follow what we write here so maybe we shouldn't point it out in that manner in the first paragraph ?

lockewritesdocs replied:
Good point. I think that we can delete the second sentence and just say:
>Starting in {es} 8.0, security is enabled and configured by default.

jkakavas commented on docs/en/getting-started/get-started-docker.asciidoc:
```diff
@@ -1,69 +1,269 @@
+You can then {kibana-ref}/docker.html[start {kib}] and enter the enrollment
+token, which is valid for 30 minutes. This token automatically applies the
+security settings from your {es} cluster, authenticates to {es} with the
+`kibana_system` user, and writes the security configuration to `kibana.yml`.
```
```suggestion
`kibana` service account, and writes the security configuration to `kibana.yml`.
```

jkakavas commented on docs/en/getting-started/get-started-docker.asciidoc:
```diff
@@ -1,69 +1,269 @@
++
+[NOTE]
+====
+Make sure that Docker is allotted at least 4GiB of memory. In Docker Desktop,
```
Is that still valid ? I run my docker containers with less with no issue

lockewritesdocs replied:
It's definitely valid for running a multi-node cluster with Docker Compose. With less memory than this, one node in my cluster continually crashed on macOS. However, this setting might not apply to the single-node cluster. If that's the case, I'll move this note to the section for using Docker Compose.

lockewritesdocs replied:
For a single-node cluster, I can run both the Elasticsearch container and the Kibana container with only 1.5 GB of memory, provided that no other containers are running. If I have another container running (like a local doc build) and try to start both Elasticsearch and Kibana, Docker stops one of those containers to regain memory.

jkakavas commented on docs/en/getting-started/get-started-docker.asciidoc:
```diff
@@ -1,69 +1,269 @@
+testing.
+
+. Start {es} in Docker. A password is generated for the `elastic` user and
+output to the terminal, plus an enrollment token for enrolling {kib}.
```
we generate a token for es nodes too

lockewritesdocs replied:
++ I'll change this to read:
>A password is generated for the `elastic` user and output to the terminal, plus enrollment tokens for enrolling {kib} and adding additional nodes to your cluster.

jkakavas commented on docs/en/getting-started/get-started-docker.asciidoc:
```diff
@@ -1,69 +1,269 @@
+* The Transport Layer Security (TLS) configuration settings are written to
+`elasticsearch.yml`.
+* A password is generated for the `elastic` user.
+* An enrollment token is generated for {kib}.
```
```suggestion
* An enrollment token is generated for {kib}.
* An enrollment token is generated for other {es} nodes.
```

jkakavas commented on docs/en/getting-started/get-started-docker.asciidoc:
```diff
@@ -1,69 +1,269 @@
+ifeval::["{release-state}"!="unreleased"]
+[source,sh,subs="attributes"]
+----
+docker run --name es-node01 --net elastic -p 9200:9200 -it docker.elastic.co/elasticsearch/elasticsearch:{version}
```
if we don't expose 9300, then no other node can join this cluster. I'm not sure we want to do this, but if we do we should point it out at least.

bytebilly replied:
Actually, other nodes running on the same host can join the cluster. I'm not sure if we want to support multi-node from multiple hosts running docker in the getting started experience, as I assume that if you run with docker, you use the same host (but I may be wrong).

jkakavas commented on docs/en/getting-started/get-started-docker.asciidoc:
```diff
@@ -1,69 +1,269 @@
+----
+`<timestamp>`:: The timestamp of when the auto-configuration process created
+the security files directory in your Docker container. Use the Docker CLI to
+locate the timestamp for this directory.
```
We will see about changing this to be a standard dir name, but for now we can add the command for the user's reference:
```
docker exec -ti es-node01 /bin/bash -c "find /usr/share/elasticsearch -name http_ca.crt"
``

lockewritesdocs replied:
Fancy! Thanks for this command; I'll definitely incorporate it into the workflow.

debadair commented on docs/en/getting-started/get-started-docker.asciidoc:
```diff
@@ -1,75 +1,300 @@
+location. These values are shown only when you start {es} for the first time.
 +
-["source","sh",subs="attributes"]
+[NOTE]
```
I think I'd pull this info about resetting the elastic password out of the procedure into a separate subsection.

lockewritesdocs replied:
I added a new section for generating passwords and enrollment tokens and moved the relevant information there. I'll probably make this change in the Elasticsearch docs as well.

Preview link: https://stack-docs_1888.docs-preview.app.elstc.co/guide/en/elastic-stack-get-started/master/get-started-stack-docker.html#docker-generate

jkakavas commented on docs/en/getting-started/get-started-docker.asciidoc:
```diff
@@ -1,75 +1,298 @@
-you can use Docker Compose.
+You can also <<get-started-docker-tls,start the {stack} with Docker Compose>> to
+create a secured, multi-node cluster with a connected {kib} instance. This
+option requires more manual steps, but results in a more resilient cluster with
```
we're trying to make it a single step too, so maybe we don't need defensive language here anymore ?

lockewritesdocs replied:
I'll update this language in the subsequent PR that introduces the simplified Docker Compose configuration.

jkakavas commented on docs/en/getting-started/get-started-docker.asciidoc:
```diff
@@ -1,75 +1,298 @@
+to the `http_ca.crt` certificate, which you'll use in the next step.
 +
-["source","sh",subs="attributes"]
+[source,sh]
+----
+docker exec -it es-node01 /bin/bash -c "find /usr/share/elasticsearch -name http_ca.crt"
+----
+
+. Copy the `http_ca.crt` security certificate from your Docker container to
+your local machine. 
++
+[source,sh]
+----
+docker cp es-node01:/usr/share/elasticsearch/config/tls_auto_config_<timestamp>/http_ca.crt .
+----
+`<timestamp>`:: The timestamp of when the auto-configuration process created
+the security files directory in your Docker container.
+
```
we will update that as we remove the timestamp from the dirname, but I haven't had the time to do this change yet. I'll make sure I'll adjust docs as part of that change

lockewritesdocs replied:
We'll need to make that change across the docs where it's used, but will be an easy change.

jkakavas reviewed (approved):
LGTM
